File: src/scala/option.scala
object MyOption {

  def main(args: Array[String]): Unit = {

    val scala = "scala"
    val java = "java"
    val php = "php"
    val rust = "rust"

    case class MeetupGuest(
                            id: Int,
                            name: String,
                            age: Int,
                            techno: List[String],
                            bestFriend: Option[Int] = None)

    val guests = List(
      MeetupGuest(1, "Rocky", 38, List(scala, java), Some(2)),
      MeetupGuest(2, "Apollo", 42, List(rust, php, java), None),
      MeetupGuest(3, "Adrienne", 34, List(php, rust), Some(1)),
      MeetupGuest(4, "Drago", 26, List(rust), None),
      MeetupGuest(5, "Mary Anne", 32, List(java, php), Some(2))
    )

    def displayAge(user: MeetupGuest): MeetupGuest = {
      println(s"${user.name} is ${user.age}")
      user
    }

    def displayUser(user: MeetupGuest): MeetupGuest = {
      println(s"Welcome ${user.name} !")
      user
    }

    def findMeetupGuest(userId: Int): Option[MeetupGuest] = guests.find(_.id == userId).headOption

    def findBestFriend(user: MeetupGuest): Option[MeetupGuest] = user.bestFriend.flatMap(findMeetupGuest)

    val userId = 1

    val lastBestFriend =
      findMeetupGuest(userId)
        .map(displayAge) // Rocky is 38
        .map(displayUser) // Welcome Rocky !
        .flatMap(findBestFriend)
        .map(displayAge) // Appolo is 42
        .map(displayUser) // Welcome Appolo !
        .flatMap(findBestFriend) // None
        .map(displayAge) // not computed
        .map(displayUser) // not computed
        .getOrElse(MeetupGuest(0, "Empty user", 0, Nil))

    displayUser(lastBestFriend);

    println("\n")
  }
}

File: src/php/option.php
<?php

function ln()
{
    echo "\n";
}

class MeetupGuest
{
    function __construct($id, $name, $age, array $techno, $bestFriend = null)
    {
        $this->id = $id;
        $this->name = $name;
        $this->age = $age;
        $this->techno = $techno;
        $this->bestFriend = $bestFriend;
    }

    function printInfo($printType)
    {
        switch ($printType) {
            case 'printAge':
                echo sprintf('%s is %d', $this->name, $this->age);
                break;
            case 'printName':
                echo sprintf('Welcome %s', $this->name);
                break;
            case 'printTechno':
                echo implode('', $this->techno);
                break;
            default:
                echo print('%s is invalid');
        }
    }
}

$scala = "scala";
$java = "java";
$php = "php";
$rust = "rust";

$guests = [
    new MeetupGuest(1, "Rocky", 38, [$scala, $java], 2),
    new MeetupGuest(2, "Apollo", 42, [$rust, $php, $java], null),
    new MeetupGuest(3, "Adrienne", 34, [$php, $rust], 1),
    new MeetupGuest(4, "Drago", 26, [$rust], null),
    new MeetupGuest(5, "Mary Anne", 32, [$java, $php], 2)
];

function findMeetupGuest($id, array $guests)
{
    $askedGuests = array_filter($guests, function (MeetupGuest $guest) use ($id) {
        return $id === $guest->id;
    });

    if (!count($askedGuests)) {
        return null;
    }

    return current($askedGuests);
}

function findBestFriend(MeetupGuest $user, array $guests)
{
    return findMeetupGuest($user->bestFriend, $guests);
}

function processBestFriend(MeetupGuest $guest, $guests)
{
    if (!$theBestFriend = findBestFriend($guest, $guests)) {
        echo sprintf("%s is sad and alone", $guest->name);
        return;
    }

    echo $theBestFriend->printInfo('printName');
    ln();
    echo $theBestFriend->printInfo('printAge');
    ln();

    return processBestFriend($theBestFriend, $guests);
}

processBestFriend($guests[0], $guests);
ln();
